### DISCUSSION TITLE
How to manually build Emacs+ (outside Homebrew)?

### DISCUSSION THREAD

**Discussion by cyrus-and** [d12frosted/homebrew-emacs-plus#739] [Q&A] [answered]
So this might sound like a weird request... but first some background, I was a pretty much happy Homebrew user, until they decided to stop supporting my old (but yet still perfectly fine) MacBook running macOS 11.7.10. Using Homebrew became a pain, every formula needed to be rebuilt from scrach, installation/updates took ages. Long story short, I switched to [Nix](https://nixos.org/) as a package manager. They also have [Emacs](https://search.nixos.org/packages?channel=24.05&show=emacs&from=0&size=50&sort=relevance&type=packages&query=emacs) in the repository, so I installed it.

Apparently, this is a vanilla Emacs which works fine, but it's still lacking when compared to Emacs+, I notice some performance issues and some flickering, and there are some difference in behaviour. For the sake of this discussion I will focus on a precise missing feature as a mean for understanding what's different.

That is, in this vanilla Emacs, setting the color of `internal-border` [has no effect](https://mail.gnu.org/archive/html/bug-gnu-emacs/2019-10/msg01843.html), while it works perfectly with Emacs+. So I tried to build the vanilla Emacs version applying the [patches](https://github.com/d12frosted/homebrew-emacs-plus/tree/master/patches/emacs-29) in this repo, but no luck. (I can share the `.nix` file if someone's interested.) The value of `system-configuration-options` is the following:

```
--prefix=/nix/store/z1fh3gda8x8hbj0v78i6h19vr9zgd16b-my-emacs-29.4
--disable-build-details
--with-modules
--disable-ns-self-contained
--with-ns
--with-compress-install
--with-toolkit-scroll-bars
--with-native-compilation
--without-imagemagick
--with-mailutils
--without-small-ja-dic
--with-tree-sitter
--without-xinput2
--without-xwidgets
--without-dbus
--without-selinux
```

So the question is, what are some other fundamental changes beside patches? What can I do/research to manage to build Emacs+ without Homebrew?

**Comment by d12frosted** [comment]
Oh. That's a shame brew dropped support of your system. 

In short, Emacs+ is just GNU Emacs with very few additions. Patches you mentioned is everything we do and nothing more. The differences might be related to Emacs version, what Emacs+ version are you using?

BTW, I have two alternatives:

- https://github.com/cmacrae/emacs; which in my memory worked well (though I don't know about internal border)
- nix-darwin has a brew module, so you can still install Emacs+ and manage other packages with nix (I do it)

**Reply by cyrus-and** [reply]
> In short, Emacs+ is just GNU Emacs with very few additions. Patches you mentioned is everything we do and nothing more. The differences might be related to Emacs version, what Emacs+ version are you using?

Oh that's super weird then. The latest Emacs+ version that I tested was 29.1, while I managed to build vanilla Emacs 29.4 from Nix + patches, here's the `.nix` file:

```nix
with (import <nixpkgs> {});

(emacs.override {}).overrideAttrs (old : {

  pname = "my-emacs";
  version = "29.4";

  src = pkgs.fetchFromGitHub {
    owner = "emacs-mirror";
    repo = "emacs";
    rev = "emacs-29.4";
    sha256 = "sha256-FCP6ySkN9mAdp2T09n6foS2OciqZXc/54guRZ0B4Z2s=";
  };

  patches = [
    (pkgs.fetchpatch {
      url = "https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/refs/heads/master/patches/emacs-28/fix-window-role.patch";
      hash = "sha256-+z/KfsBm1lvZTZNiMbxzXQGRTjkCFO4QPlEK35upjsE=";
    })
    (pkgs.fetchpatch {
      url = "https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/refs/heads/master/patches/emacs-28/no-frame-refocus-cocoa.patch";
      hash = "sha256-QLGplGoRpM4qgrIAJIbVJJsa4xj34axwT3LiWt++j/c=";
    })
    (pkgs.fetchpatch {
      url = "https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/refs/heads/master/patches/emacs-29/poll.patch";
      hash = "sha256-jN9MlD8/ZrnLuP2/HUXXEVVd6A+aRZNYFdZF8ReJGfY=";
    })
    (pkgs.fetchpatch {
      url = "https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/refs/heads/master/patches/emacs-29/round-undecorated-frame.patch";
      hash = "sha256-uYIxNTyfbprx5mCqMNFVrBcLeo+8e21qmBE3lpcnd+4=";
    })
    (pkgs.fetchpatch {
      url = "https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/refs/heads/master/patches/emacs-28/system-appearance.patch";
      hash = "sha256-oM6fXdXCWVcBnNrzXmF0ZMdp8j0pzkLE66WteeCutv8=";
    })
  ];
})
```

> BTW, I have two alternatives:

Awesome! <3 I will look into that.

But at this point I have to admit I'm also curious as what's happening here...

**Reply by cyrus-and** [reply, marked as answer]
So apparently the culprit is just Nix, I suspect it's the Nix's `AppKit` dependency that differs somehow from the one present in my system.

Building Emacs manually in this way (even without patches) also produces an apparently much snappier and stable release.

Still waiting for someone to explain to me why is that. See also: https://discourse.nixos.org/t/building-emacs-with-nix-produce-a-different-lacking-version-when-build-manually-on-macos/54966
